File: knowledge-graph-service/app/services/embedding_service.py
from openai import OpenAI
from app.core.config import get_settings
import tiktoken
import logging
from typing import Tuple, List

logger = logging.getLogger(__name__)
settings = get_settings()


class EmbeddingService:
    """OpenAI 임베딩 생성"""

    def __init__(self):
        # OpenAI 클라이언트 초기화
        if settings.openai_base_url:
            # GMS 서비스 사용
            self.client = OpenAI(
                api_key=settings.openai_api_key, base_url=settings.openai_base_url
            )
            logger.debug(f"✅ OpenAI 클라이언트 (GMS): {settings.openai_base_url}")
        else:
            # 기본 OpenAI API 사용
            self.client = OpenAI(api_key=settings.openai_api_key)
            logger.debug("✅ OpenAI 클라이언트 (Base API)")

        self.model = settings.openai_model
        self.encoding = tiktoken.encoding_for_model("gpt-3.5-turbo")

    def count_tokens(self, text: str) -> int:
        """
        토큰 수 계산

        Args:
            text: 계산할 텍스트

        Returns:
            토큰 개수
        """
        return len(self.encoding.encode(text))

    def generate_embedding(self, text: str) -> Tuple[List[float], int]:
        """
        임베딩 생성 (전체 내용, 제한 없음)

        Args:
            text: 임베딩할 텍스트 (전체 노트 내용)

        Returns:
            (임베딩 벡터, 토큰 개수)

        Raises:
            Exception: OpenAI API 호출 실패
        """
        try:
            # 1. 토큰 수 계산 (정보용)
            token_count = self.count_tokens(text)
            logger.debug(f"📊 토큰 수: {token_count}개")

            # 2. OpenAI API 호출 (제한 없이 전체 임베딩)
            logger.debug(f"🤖 임베딩 생성 중...")

            response = self.client.embeddings.create(
                model=self.model,
                input=text,  # 👈 전체 내용 그대로
                encoding_format="float",
            )

            # 3. 임베딩 추출
            embedding = response.data[0].embedding

            logger.debug(f"✅ 임베딩 생성 완료: {len(embedding)}차원")

            return embedding, token_count

        except Exception as e:
            logger.error(f"❌ 임베딩 생성 실패: {e}")
            raise


# 싱글톤 인스턴스
embedding_service = EmbeddingService()

File: backend/secondbrain/src/main/java/uknowklp/secondbrain/api/note/service/EmbeddingService.java
package uknowklp.secondbrain.api.note.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import com.openai.client.OpenAIClient;
import com.openai.client.okhttp.OpenAIOkHttpClient;
import com.openai.models.CreateEmbeddingResponse;
import com.openai.models.EmbeddingCreateParams;

import jakarta.annotation.PostConstruct;
import lombok.extern.slf4j.Slf4j;

// GMS의 text-embedding-3-small 모델을 사용하여 검색어를 1536차원 벡터로 변환
@Slf4j
@Service
public class EmbeddingService {

	@Value("${gms.api-key}")
	private String apiKey;

	@Value("${gms.embedding-model}")
	private String embeddingModel;

	@Value("${gms.openai-base-url}")
	private String openaiBaseUrl;

	private OpenAIClient client;

	@PostConstruct
	public void init() {
		// FastAPI와 동일: OpenAI SDK 초기화 (base_url 설정)
		this.client = OpenAIOkHttpClient.builder()
			.apiKey(apiKey)
			.baseUrl(openaiBaseUrl)
			.build();

		log.debug("✅ OpenAI 클라이언트 초기화 (GMS): {}", openaiBaseUrl);
	}

	// 텍스트를 1536차원 임베딩 벡터로 변환 (OpenAI SDK 사용)
	public List<Double> generateEmbedding(String text) {
		try {
			log.debug("🤖 임베딩 생성 중 - 모델: {}, 텍스트 길이: {}자", embeddingModel, text.length());

			// FastAPI와 동일: client.embeddings.create()
			EmbeddingCreateParams params = EmbeddingCreateParams.builder()
				.model(embeddingModel)
				.input(EmbeddingCreateParams.Input.ofString(text))
				.encodingFormat(EmbeddingCreateParams.EncodingFormat.FLOAT)
				.build();

			CreateEmbeddingResponse response = client.embeddings().create(params);

			// 임베딩 벡터 추출
			List<Double> embedding = response.data().get(0).embedding();

			log.debug("✅ 임베딩 생성 완료 - 차원: {}", embedding.size());

			return embedding;

		} catch (Exception e) {
			log.error("❌ 임베딩 생성 실패: {}", e.getMessage(), e);
			throw new RuntimeException("임베딩 생성 중 오류 발생", e);
		}
	}
}
